src/equations/acoustic_perturbation_2d.jl

```diff
@@ -116,7 +116,9 @@ function initial_condition_constant(x, t, equations::AcousticPerturbationEquatio
     v2_prime = 0.0
     p_prime_scaled = 0.0
 
-    return SVector(v1_prime, v2_prime, p_prime_scaled, global_mean_vars(equations)...)
+    return SVector(map(v -> convert(eltype(x), v),
+                       (v1_prime, v2_prime, p_prime_scaled,
+                        global_mean_vars(equations)...)))
```

Comment:
This is one of the more complex cases since several input types determine the output type. In general, we would like this to be as generic as possible. For example, we would like to have the option to use AD through it, e.g., to differentiate with respect to the `global_mean_vars(equations)` as in https://trixi-framework.github.io/Trixi.jl/stable/tutorials/differentiable_programming/#Computing-other-derivatives
In this specific case, we can use that the `SVector` constructor promotes its arguments. Thus, I would use something like

```julia
    # We want to make this function as general as possible. Since the
    # `SVector` constructor below promotes its arguments, we just use
    # integers for the zero values.
    v1_prime = 0
    v2_prime = 0
    p_prime_scaled = 0

    return SVector(v1_prime, v2_prime, p_prime_scaled, global_mean_vars(equations)...)
```

Alternatively, we could use something like

```julia
    RealT = eltype(x)
    v1_prime = zero(RealT)
    v2_prime = zero(RealT)
    p_prime_scaled = zero(RealT)

    return SVector(v1_prime, v2_prime, p_prime_scaled, global_mean_vars(equations)...)
```

I don't have a very strong opinion on this but I think the second version may be easier to generalize to other cases where non-integer values are present.

Reply:
```This is one of the more complex cases since several input types determine the output type.```
But should we base the assumption that all inputs are restricted to only one type like solely `Float32` or `Float64`?

Reply:
I would like to avoid the assumption that we have to deal with only one type. As in the case linked above, we also have to deal with AD involving an additional type. If we touch the code right now, we should keep this in mind.

Reply:
What I am supposed to achieve in my project is something like
1. Convert all relevant inputs to `Float32` (current default branch is using `Float32` and we ignore `Float64` for now) when copy to GPU.
2. Keep all data running in `Float32` as much as possible.
3. Convert all output back to its original type (whatever it is,`Float32` or `Float64`) when copy back to CPU.

This PR aims to prevent the accidental promotion from `Float32` to `Float64` when calling certain functions from the repository during step 2 (as mentioned above). Why can't we assume that everything is already in `Float32`? I just want to make sure we are on the same page and I would like to understand why this assumption fails.

Reply:
Some people are using automatic/algorithmic differentiation (AD) with Trixi.jl, e.g., `@DanielDoehring` in https://arxiv.org/abs/2403.05144. If you use AD, `global_mean_vars(equations)` may return some dual numbers that cannot be converted to `Float32`.